### PR TITLE
metrics: Expose xfrm stats in prometheus

### LINKED ISSUE #14725: Expose xfrm stats as Prometheus metrics for IPSec
When IPSec is enabled, we should expose [xfrm errors](https://www.kernel.org/doc/Documentation/networking/xfrm_proc.txt) as Prometheus metrics. If those counters are increasing, they can be indications of misconfigured keys, corrupted data, etc.

### PR DESCRIPTION
**Pull request by sayboras**
### Description

This commit is to expose xfrm stats via prometheus metrics if IPSec is
enabled.

Fixes: #14725
Signed-off-by: Tam Mach <tam.mach@isovalent.com>

### Testing

Testing was done locally with kind clusters, please find below details:

<details>
<summary>ip sec is NOT enabled</summary>

```
root@kind-control-plane:/home/cilium# cilium metrics list -p ipsec
Metric   Labels   Value
```

</details>

<details>
<summary>ip sec is enabled</summary>

```

root@kind-worker:/home/cilium# cilium metrics list -p ipsec

cilium_ipsec_xfrm_error   error="acquire" type="inbound"               0.000000
cilium_ipsec_xfrm_error   error="bundle_check" type="outbound"         0.000000
cilium_ipsec_xfrm_error   error="bundle_generation" type="outbound"    0.000000
cilium_ipsec_xfrm_error   error="forward_header" type="inbound"        0.000000
cilium_ipsec_xfrm_error   error="header" type="inbound"                0.000000
cilium_ipsec_xfrm_error   error="no_buffer" type="inbound"             0.000000
cilium_ipsec_xfrm_error   error="no_policy" type="inbound"             0.000000
cilium_ipsec_xfrm_error   error="no_state" type="inbound"              0.000000
cilium_ipsec_xfrm_error   error="no_state" type="outbound"             0.000000
cilium_ipsec_xfrm_error   error="other" type="inbound"                 0.000000
cilium_ipsec_xfrm_error   error="other" type="outbound"                0.000000
cilium_ipsec_xfrm_error   error="policy" type="inbound"                0.000000
cilium_ipsec_xfrm_error   error="policy" type="outbound"               0.000000
cilium_ipsec_xfrm_error   error="policy_blocked" type="inbound"        0.000000
cilium_ipsec_xfrm_error   error="policy_blocked" type="outbound"       0.000000
cilium_ipsec_xfrm_error   error="policy_dead" type="outbound"          0.000000
cilium_ipsec_xfrm_error   error="state_expired" type="inbound"         0.000000
cilium_ipsec_xfrm_error   error="state_expired" type="outbound"        0.000000
cilium_ipsec_xfrm_error   error="state_invalid" type="inbound"         0.000000
cilium_ipsec_xfrm_error   error="state_invalid" type="outbound"        0.000000
cilium_ipsec_xfrm_error   error="state_mismatched" type="inbound"      0.000000
cilium_ipsec_xfrm_error   error="state_mode" type="inbound"            0.000000
cilium_ipsec_xfrm_error   error="state_mode" type="outbound"           0.000000
cilium_ipsec_xfrm_error   error="state_protocol" type="inbound"        0.000000
cilium_ipsec_xfrm_error   error="state_protocol" type="outbound"       0.000000
cilium_ipsec_xfrm_error   error="state_sequence" type="inbound"        0.000000
cilium_ipsec_xfrm_error   error="state_sequence" type="outbound"       0.000000
cilium_ipsec_xfrm_error   error="template_mismatched" type="inbound"   0.000000

root@kind-control-plane:/home/cilium# cat /proc/net/xfrm_stat 
XfrmInError                     0
XfrmInBufferError               0
XfrmInHdrError                  0
XfrmInNoStates                  0
XfrmInStateProtoError           0
XfrmInStateModeError            0
XfrmInStateSeqError             0
XfrmInStateExpired              0
XfrmInStateMismatch             0
XfrmInStateInvalid              0
XfrmInTmplMismatch              0
XfrmInNoPols                    0
XfrmInPolBlock                  0
XfrmInPolError                  0
XfrmOutError                    0
XfrmOutBundleGenError           0
XfrmOutBundleCheckError         0
XfrmOutNoStates                 0
XfrmOutStateProtoError          0
XfrmOutStateModeError           0
XfrmOutStateSeqError            0
XfrmOutStateExpired             0
XfrmOutPolBlock                 0
XfrmOutPolDead                  0
XfrmOutPolError                 0
XfrmFwdHdrError                 0
XfrmOutStateInvalid             0
XfrmAcquireError                0

```

</details>

```release-note
metrics: Expose xfrm stats in prometheus metrics
```

### CODE REVIEW

File: pkg/datapath/linux/ipsec/xfrm_collector.go [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+
+func newXFRMCollector(statFn func() (procfs.XfrmStat, error)) prometheus.Collector {
+	return &xfrmCollector{
+		xfrmStatFunc: statFn,
+
+		// Inbound errors
+		xfrmInErrorDesc: prometheus.NewDesc(
+			prometheus.BuildFQName(metrics.Namespace, subsystem, "xfrm_in_error"),
+			"All inbound errors which is not matched others",
+			nil, map[string]string{labelErrorType: labelErrorTypeInbound},
```
Comment (sayboras) [inline pkg/datapath/linux/ipsec/xfrm_collector.go]: I am thinking about about adding labels for cluster name and node name, but existing ebpf related metrics don't have these labels :man_shrugging:

File: pkg/datapath/linux/ipsec/xfrm_collector.go [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+}
+
+func (x *xfrmCollector) Collect(ch chan<- prometheus.Metric) {
+	stats, err := x.xfrmStatFunc()
+	if err != nil {
+		log.WithError(err).Error("Error while getting xfrm stats")
+		return
+	}
+
+	ch <- prometheus.MustNewConstMetric(x.xfrmInErrorDesc, prometheus.GaugeValue, float64(stats.XfrmInError))
```
Comment (sayboras) [inline pkg/datapath/linux/ipsec/xfrm_collector.go]: few of metrics can be combined with diff tag values, but it is not really help much in term of storage (as real number of metric is all combination of tag values), so just to simplicity, I keep it 1 to 1 mapping between metric and value in /proc/net/xfrm file.

File: pkg/datapath/linux/node.go [hunk 3]
```diff
@@ -1134,6 +1141,9 @@ func (n *linuxNodeHandler) NodeDelete(oldNode nodeTypes.Node) error {
 // Must be called with linuxNodeHandler.mutex held.
 func (n *linuxNodeHandler) nodeDelete(oldNode *nodeTypes.Node) error {
 	if oldNode.IsLocal() {
+		if n.nodeConfig.EnableIPSec {
+			metrics.Unregister(n.ipsecMetricCollector)
```
Comment (pchaigno) [inline pkg/datapath/linux/node.go:1145]: Very Prometheus-naive question:
Can registering conditionally on IPsec being enabled have any lasting consequences if someone temporarily disabled IPsec? That is, can unregistering cause past data to be lost, or is the data persisted in some other way?

**Review by pchaigno — APPROVED** [review]
Thanks for taking care of this! :pray: :tada: 

Ok on IPsec-specific bits :+1: I'm too unfamiliar with Prometheus to approve the rest.

File: pkg/datapath/linux/node.go (hunk 3 above)
Comment (sayboras) [reply pkg/datapath/linux/node.go:1145]: AFAIK, cilium only stores metrics in memory, and didn't actually push to any external consumer. Consumers can get metric values either via cli or /metrics endpoint.

Coming back to your question, if ipsec is disabled, I think we should stop exposing any related metrics at all, as these values are no longer making sense in current setup. Additionally, this is one of the reason I configure them as Gauge type.

File: pkg/datapath/linux/ipsec/xfrm_collector.go [hunk 4]
```diff
@@ -0,0 +1,277 @@
+// SPDX-License-Identifier: Apache-2.0
```
Comment (aanm) [inline pkg/datapath/linux/ipsec/xfrm_collector.go:1]: please add this file as a part of the CODEOWNERS file for the `metrics` team

File: pkg/datapath/linux/ipsec/xfrm_collector.go [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+	xfrmOutBundleCheckErrorDesc *prometheus.Desc
+	xfrmOutNoStatesDesc         *prometheus.Desc
+	xfrmOutStateProtoErrorDesc  *prometheus.Desc
+	xfrmOutStateModeErrorDesc   *prometheus.Desc
+	xfrmOutStateSeqErrorDesc    *prometheus.Desc
+	xfrmOutStateExpiredDesc     *prometheus.Desc
+	xfrmOutPolBlockDesc         *prometheus.Desc
+	xfrmOutPolDeadDesc          *prometheus.Desc
+	xfrmOutPolErrorDesc         *prometheus.Desc
+	xfrmOutStateInvalidDesc     *prometheus.Desc
```
Comment (aanm) [inline pkg/datapath/linux/ipsec/xfrm_collector.go]: These are a lot of metrics :exploding_head: Was it considered having `In` and `Out` and all the error types as labels of one single metric instead?

**Review by aanm — CHANGES_REQUESTED** [review]

File: pkg/datapath/linux/ipsec/xfrm_collector.go (hunk 5 above)
Comment (sayboras) [reply pkg/datapath/linux/ipsec/xfrm_collector.go]: making sense to me, I have collated all into one single metric cilium_ipsec_xfrm_error{error=reason,type=inbound|outbound}

**Comment by sayboras** [comment]
force push to address the above comment, also update related docs.

**Review by qmonnet — APPROVED** [review]

**Review by pchaigno — APPROVED** [review]

**Review by joamaki — APPROVED** [review]

**Review by aanm — APPROVED** [review]
:tada:

**Comment by pchaigno** [comment]
The code changes are only covered in the new unit test, which is passing. Smoke tests are also passing and reviews are in. Marking ready to merge.

**Review by michi-covalent — APPROVED** [review]

**Review by christarazi — APPROVED** [review]
:rocket:

**Pull request merged**